### PR TITLE
mv fifo over from ip-contributions

### PR DESCRIPTION
**Pull request by schoeberl**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

   - new feature/API                 

#### API Impact

ad to the future Chisel std lib (mv from ip-contributions)

#### Desired Merge Strategy

  - Squash: The PR will be squashed and merged (choose this if you have no preference. 

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by schoeberl** [comment]
Because it delivers nice implementation variations all out of an abstract class. There you can really change the implementation type without changing the client code. I don't think this can easily be done with Queue.

File: integration-tests/src/test/scala/chisel3/std/fifo/FifoSpec.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
+// Author: Martin Schoeberl (martin@jopdesign.com)
+// License: this code is released into the public domain, see README.md and http://unlicense.org/
```
Comment (mwachs5) [inline integration-tests/src/test/scala/chisel3/std/fifo/FifoSpec.scala:2]: As you are the author of this code and have to sign the CLA agreement to contribute to the repo, it seems we would want to keep the same license for ip-contributions as the rest of the repo. Thoughts?

File: stdlib/src/main/scala/chisel3/std/fifo/CombFifo.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Author: Martin Schoeberl (martin@jopdesign.com)
+// License: this code is released into the public domain, see README.md and http://unlicense.org/
+
+package chisel3.std.fifo
+
+import chisel3._
+
+/**
+  * A FIFO queue combining the memory FIFO with a single stage double buffer FIFO
```
Comment (mwachs5) [inline stdlib/src/main/scala/chisel3/std/fifo/CombFifo.scala:9]: the scaladoc in these is a little bit mis-formatted (suprised scalafmt is not flagging it?). I think we should have a really high bar for documentation for anything in the stdlib

File: integration-tests/src/test/scala/chisel3/std/fifo/FifoSpec.scala (hunk 1 above)
Comment (schoeberl) [reply integration-tests/src/test/scala/chisel3/std/fifo/FifoSpec.scala:2]: Yes, I am fine with the Chisel license. This was just the license used in ip-contribution. Will change it.

**Comment by schoeberl** [comment]
There was a decision to keep those FIFOs in ip-contribution.

**Pull request closed without merging**